rcutils logging macro warning

I was using the new logging macros like this:

```c
#include "rcutils/logging_macros.h"
// ...
    RCUTILS_LOG_ERROR(
      "failed to clean up on error (leaking memory): '%s'", rcutils_get_error_string_safe());
```

But I get a warning like this:

```
In file included from /Users/william/ros2_ws/src/ros2/rcutils/src/split.c:26:
/Users/william/ros2_ws/build/rcutils/include/rcutils/logging_macros.h:948:13: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
    format, ## __VA_ARGS__)
            ^
/Users/william/ros2_ws/build/rcutils/include/rcutils/logging_macros.h:93:68: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
      rcl_log(&__rcutils_logging_location, severity, name, format, ## __VA_ARGS__); \
                                                                   ^
2 warnings generated.
```

I looked into it briefly, but I didn't dig too deep. Am I doing something wrong or is this just a warning we should fix in the logging macros?

The warning only references the line on which the header is included, but if I remove all uses of the logging macro (but leave the include) the warning goes away.

Unfortunately this fix forces users to have a variadic argument making it impossible to do
`RCUTILS_LOG_ERROR("I have no argument");` so we either need a way to handle empty va_args list or use a different logic on a per platform basis.
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2738)](http://ci.ros2.org/job/ci_linux/2738/)

```
In file included from /home/mikael/work/ros2/ros2_ws/src/ros2/rcl/rcl/src/rcl/common.c:26:0:
/home/mikael/work/ros2/ros2_ws/src/ros2/rcl/rcl/src/rcl/common.c: In function ‘rcl_impl_getenv’:
/home/mikael/work/ros2/ros2_ws/install_debug_isolated/rcutils/include/rcutils/logging_macros.h:93:79: error: expected expression before ‘)’ token
       rcl_log(&__rcutils_logging_location, severity, name, format, __VA_ARGS__); \
                                                                               ^
/home/mikael/work/ros2/ros2_ws/install_debug_isolated/rcutils/include/rcutils/logging_macros.h:945:3: note: in expansion of macro ‘RCUTILS_LOG_COND_NAMED’
   RCUTILS_LOG_COND_NAMED( \
   ^
/home/mikael/work/ros2/ros2_ws/src/ros2/rcl/rcl/src/rcl/common.c:37:3: note: in expansion of macro ‘RCUTILS_LOG_ERROR’
   RCUTILS_LOG_ERROR("No variadic arg");
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -130,6 +130,12 @@ if(BUILD_TESTING)
     APPEND_LIBRARY_DIRS ${extra_lib_dirs})
   target_link_libraries(test_logging_macros ${PROJECT_NAME})
 
+  add_executable(test_logging_macros_c test/test_logging_macros.c)
+  target_link_libraries(test_logging_macros_c ${PROJECT_NAME})
+  ament_add_test(test_logging_macros_c
+    COMMAND "$<TARGET_FILE:test_logging_macros_c>"
+    GENERATE_RESULT_FOR_RETURN_CODE_ZERO)
+
   # This subdirectory extends both extra_test_libraries, extra_test_env, and extra_lib_dirs.
   add_subdirectory(test/memory_tools)
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,6 +2,15 @@ cmake_minimum_required(VERSION 3.5)
 
 project(rcutils)
 
+# Default to C11
+if(NOT CMAKE_C_STANDARD)
+  set(CMAKE_C_STANDARD 11)
+endif()
+# Default to C++14
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 14)
+endif()
+
 find_package(ament_cmake REQUIRED)
 
 include_directories(include)

--- a/resource/logging_macros.h.em
+++ b/resource/logging_macros.h.em
@@ -84,13 +84,13 @@ extern "C"
  * \param format The format string
  * \param ... The variable arguments for the format strings
  */
-#define RCUTILS_LOG_COND_NAMED(severity, condition_before, condition_after, name, format, ...) \
+#define RCUTILS_LOG_COND_NAMED(severity, condition_before, condition_after, name, ...) \
   { \
     RCUTILS_LOGGING_AUTOINIT \
     static rcl_log_location_t __rcutils_logging_location = {__func__, __FILE__, __LINE__}; \
     condition_before \
     if (severity >= g_rcl_logging_severity_threshold) { \
-      rcl_log(&__rcutils_logging_location, severity, name, format, ## __VA_ARGS__); \
+      rcl_log(&__rcutils_logging_location, severity, name, __VA_ARGS__); \
     } \
     condition_after \
   }
@@ -397,11 +397,11 @@ def get_macro_arguments(suffix):
  * \param format The format string
  * \param ... The variable arguments for the format strings
  */
-# define RCUTILS_LOG_@(severity)@(suffix)(@(get_macro_parameters(suffix))format, ...) \
+# define RCUTILS_LOG_@(severity)@(suffix)(@(get_macro_parameters(suffix))...) \
   RCUTILS_LOG_COND_NAMED( \
     RCUTILS_LOG_SEVERITY_@(severity), \
     @(get_macro_arguments(suffix))\
-    format, ## __VA_ARGS__)
+    __VA_ARGS__)
 @[ end for]@
 #endif
 ///@@}

--- a/test/test_logging_macros.c
+++ b/test/test_logging_macros.c
@@ -1,0 +1,118 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string.h>
+
+#include "rcutils/logging_macros.h"
+
+size_t g_log_calls = 0;
+
+struct LogEvent
+{
+  rcl_log_location_t * location;
+  int severity;
+  const char * name;
+  char * message;
+};
+struct LogEvent g_last_log_event;
+
+void custom_handler(
+  rcl_log_location_t * location,
+  int severity, const char * name, const char * format, va_list * args)
+{
+  g_log_calls += 1;
+  g_last_log_event.location = location;
+  g_last_log_event.severity = severity;
+  g_last_log_event.name = name ? name : "";
+  if (g_last_log_event.message) {
+    free(g_last_log_event.message);
+  }
+  const size_t size = 1024;
+  g_last_log_event.message = malloc(size);
+  vsnprintf(g_last_log_event.message, size, format, *args);
+}
+
+int main(int argc, char ** argv)
+{
+  (void)argc;
+  (void)argv;
+
+  g_log_calls = 0;
+  if (g_rcl_logging_initialized) {
+    return 1;
+  }
+  rcl_logging_initialize();
+  if (!g_rcl_logging_initialized) {
+    return 2;
+  }
+
+  rcl_logging_output_handler_t previous_output_handler = \
+    rcl_logging_get_output_handler();
+  rcl_logging_set_output_handler(custom_handler);
+
+  RCUTILS_LOG_INFO("empty message");
+  if (g_log_calls != 1u) {
+    return 3;
+  }
+  if (!g_last_log_event.location) {
+    return 4;
+  }
+  if (strcmp(g_last_log_event.location->function_name, "main")) {
+    return 5;
+  }
+  if (g_last_log_event.location->line_number != 64u) {
+    return 6;
+  }
+  if (g_last_log_event.severity != RCUTILS_LOG_SEVERITY_INFO) {
+    return 7;
+  }
+  if (strcmp(g_last_log_event.name, "")) {
+    return 8;
+  }
+  if (strcmp(g_last_log_event.message, "empty message")) {
+    return 9;
+  }
+
+  RCUTILS_LOG_INFO("message %s", "foo");
+  if (g_log_calls != 2u) {
+    return 10;
+  }
+  if (!g_last_log_event.location) {
+    return 11;
+  }
+  if (strcmp(g_last_log_event.location->function_name, "main")) {
+    return 12;
+  }
+  if (g_last_log_event.location->line_number != 87u) {
+    return 13;
+  }
+  if (g_last_log_event.severity != RCUTILS_LOG_SEVERITY_INFO) {
+    return 14;
+  }
+  if (strcmp(g_last_log_event.name, "")) {
+    return 15;
+  }
+  if (strcmp(g_last_log_event.message, "message foo")) {
+    return 16;
+  }
+
+  rcl_logging_set_output_handler(previous_output_handler);
+  if (g_last_log_event.message) {
+    free(g_last_log_event.message);
+  }
+  g_rcl_logging_initialized = false;
+  if (g_rcl_logging_initialized) {
+    return 17;
+  }
+}

